design-documents/scalars-for-templates.md

```diff
@@ -0,0 +1,177 @@
+     * 
+     * @return DirectiveProcessorInterface[]|null
+     */
+    public function getSubDirectives(): ?array;
+    
+    /**
+     * Default filters to apply if none provided in a template.
+     *
+     * @return string[]|null
+     */
+    public function getDefaultFilters(): ?array;
+}
+```
+ 
+_Magento\Framework\Filter\Template_ will receive the list of such directive processors
+via di.xml, parse all directives in given templates and give them to processors
+until there is no directives left. That means that directives' HTML may contain
+directives as well which will not be rendered by it's directive.
+ 
+Consider the example:
+```html
+<h2>{{currentStoreName}}</h2>
+<p>{{trans 'Hello, %name! %text' name=customer.name text='txt'|raw }}</p>
+{{if show}}
+<div>Showing</div>
+{{else}}
+<div>No show</div>
+{{/if}}
+```
+ 
+When processing given template the _Template_ will try to find _DirectiveProcessorInterface_ whose
+_getName()_ returns __currentStoreName__ to process the first directive, it will
+call it's _process()_ method with _(null, null, [])_ since there is no value or parameters
+provided, after it will call it's _getDefaultFilters()_ and it will return _'escape'_
+because there are no filters provided in the template. _'escape'_ filter will be applied.
+ 
+Next _Template_ will try to find _'trans'_ directive, it will pass _('Hello, %name! %text',
+null, ['name' => 'Karen Smith', 'text' => 'txt])_ to it's _process()_ method. The directive's
+_getDefaultFilters()_ will not be called since _'raw'_ filter is stated.
+ 
+Then _Template_ will find __if__ directive, it's _getSubDirectives()_ will be called
+since the directive contains HTML
+and it will return __else__ directive processor. It's _process()_ method will
+receive _(false, '<div>Showing</div>', [])_ and then the __else__ directive will
+receive _(null, '<div>No show</div>', [])_. Both __if's__ and __else's__ _getDefaultFilters()_ will return nulls.
+ 
+Another thing that 3rd party developers might want to extend/add are filters, I propose such SPI:
+```php
+interface FilterProcessorInterface
+{
+    /**
+     * Process a value.
+     *
+     * @return string
+     */
+    public function filterValue(string $value): string;
+    
+    /**
+     * This filter's unique name.
+     *
+     * @return string
+     */
+    public function getName(): string;
+}
+```
+ 
+_Template_ class would receive a list of such filters and use them accordingly.
+ 
+###### Deprecate old methods
+Old public methods following _\<Directive Name\>Directive()_ pattern must be deprecated in favor of new directive
+processors.
+
+#### Backward compatibility
+This would not be a backward compatible change - we use objects in our own
+E-mail templates in core Magento and developers are using objects in their own templates
+as well.
+
+In order to preserve backward compatibility E-mail templates read from _.html_ files will keep working
```

Comment:
What to do with templates saved to database?

Reply:
I mention this in the document below this statement - render them as they were. I was thinking about adding a flag to newly created/updated templates to render them with compliance to new restrictions